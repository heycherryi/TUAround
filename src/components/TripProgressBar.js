import React from 'react';
import PropTypes from 'prop-types';
import { ProgressBar } from 'react-bootstrap';
import { formatDistanceTraveled, isTimeLaterThanNow } from '../util';

function TripProgressBar({ trip, tripUpdates }) {
  function renderEmptyProgressBar() {
    return (
      <div className="trip-progress-bar trip-progress-bar-empty my-4">
        <ProgressBar min={0} max={100} label={''} now={0}></ProgressBar>
      </div>
    );
  }

  function renderCompleteProgressBar() {
    return (
      <div className="trip-progress-bar trip-progress-bar-completed my-4">
        <ProgressBar min={0} max={totalTripDistance} label={formatDistanceTraveled(totalTripDistance)} now={totalTripDistance}></ProgressBar>
      </div>
    );
  }

  let currentStopDistance = 0;
  let totalTripDistance = trip.stop_times[trip.stop_times.length - 1].shape_dist_traveled;

  // If no updates provided, render either empty or completed bar based on static schedule
  if (tripUpdates.length === 0 || typeof tripUpdates[0].trip_update === 'undefined') {
    if (!isTimeLaterThanNow(trip.stop_times[trip.stop_times.length - 1].arrival_time)) {
      return renderCompleteProgressBar();
    }
    // Trip hasn't completed and there's no updates; Assume it's before tracking picks it up
    return renderEmptyProgressBar();
  }

  const currentStopUpdate = tripUpdates[0].trip_update.stop_time_update.find((item) => {
    if (typeof item.departure !== 'undefined' && item.departure.time * 1000 > Date.now()) {
      return true;
    }
    if (typeof item.arrival !== 'undefined' && item.arrival.time * 1000 > Date.now()) {
      return true;
    }
    return false;
  });
  if (typeof currentStopUpdate === 'undefined') {
    return renderCompleteProgressBar();
  }
  const currentStop = trip.stop_times.find((i) => i.stop_sequence == currentStopUpdate.stop_sequence);

  // Start of trip
  if (!currentStop.shape_dist_traveled) {
    return renderEmptyProgressBar();
  }

  currentStopDistance = parseFloat(currentStop.shape_dist_traveled);

  return (
    <div className="trip-progress-bar my-4">
      <ProgressBar min={0} max={totalTripDistance} label={formatDistanceTraveled(currentStopDistance)} now={currentStopDistance}></ProgressBar>
    </div>
  );
}

TripProgressBar.propTypes = {
  trip: PropTypes.object.isRequired,
  tripUpdates: PropTypes.any,
};

TripProgressBar.defaultProps = {
  tripUpdates: [],
};

export default TripProgressBar;
