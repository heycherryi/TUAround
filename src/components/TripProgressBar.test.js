/* globals test, expect, afterEach */

import React from 'react';
import { render, screen } from '@testing-library/react';
import MockDate from 'mockdate';
import TripProgressBar from './TripProgressBar';

const tripFixture = require('../fixtures/trips-270708.json');
const tripUpdatesFixture = require('../fixtures/trip_updates.json');

afterEach(() => {
  MockDate.reset();
});

test('renders TripProgressBar', () => {
  MockDate.set('Fri Jul 22 2022 22:10:00 GMT-0500');
  const tripUpdates = tripUpdatesFixture.filter((i) => i.trip_update.trip.trip_id === tripFixture.trip_gid);
  const { container } = render(
    <TripProgressBar trip={tripFixture} tripUpdates={tripUpdates} />,
  );
  expect(container.querySelector('.trip-progress-bar > .progress > div')).toHaveStyle({ width: '17.551%' });
  expect(screen.getByText('2.44 mi')).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test('renders TripProgressBar for completed trip', () => {
  MockDate.set('Fri Jul 22 2022 23:59:00 GMT-0500');
  const { container } = render(
    <TripProgressBar trip={tripFixture} tripUpdates={[]} />,
  );
  expect(container.querySelector('.trip-progress-bar-completed')).toBeInTheDocument();
  expect(container.querySelector('.trip-progress-bar > .progress > div')).toHaveStyle({ width: '100%' });
  expect(screen.getByText('13.92 mi')).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test('renders empty TripProgressBar for trip not yet started', () => {
  MockDate.set('Fri Jul 22 2022 10:45:00 GMT-0500');
  const { container } = render(
    <TripProgressBar trip={tripFixture} tripUpdates={[]} />,
  );
  expect(container.querySelector('.trip-progress-bar-empty')).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});
