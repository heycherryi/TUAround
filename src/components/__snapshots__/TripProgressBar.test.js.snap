// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TripProgressBar 1`] = `
<div>
  <div
    class="trip-progress-bar my-4"
  >
    <div
      class="progress"
    >
      <div
        aria-valuemax="22.3979"
        aria-valuemin="0"
        aria-valuenow="3.9311"
        class="progress-bar"
        role="progressbar"
        style="width: 17.551%;"
      >
        2.44 mi
      </div>
    </div>
  </div>
</div>
`;

exports[`renders TripProgressBar for completed trip 1`] = `
<div>
  <div
    class="trip-progress-bar trip-progress-bar-completed my-4"
  >
    <div
      class="progress"
    >
      <div
        aria-valuemax="22.3979"
        aria-valuemin="0"
        aria-valuenow="22.3979"
        class="progress-bar"
        role="progressbar"
        style="width: 100%;"
      >
        13.92 mi
      </div>
    </div>
  </div>
</div>
`;

exports[`renders empty TripProgressBar for trip not yet started 1`] = `
<div>
  <div
    class="trip-progress-bar trip-progress-bar-empty my-4"
  >
    <div
      class="progress"
    >
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="0"
        class="progress-bar"
        role="progressbar"
        style="width: 0%;"
      />
    </div>
  </div>
</div>
`;
